### PR TITLE
CBG-1542 - Command line flags for persistent config

### PR DESCRIPTION
**Pull request by IsaacLambat**
- [x] http://uberjenkins.sc.couchbase.com:8080/job/sync-gateway-integration/963
- [x] http://uberjenkins.sc.couchbase.com:8080/job/sync-gateway-integration/995/
- [x] http://uberjenkins.sc.couchbase.com:8080/job/sync-gateway-integration/1009/
- [x] http://uberjenkins.sc.couchbase.com:8080/job/sync-gateway-integration/1011/

- Errors are returned if there is problems parsing the CLI flags (eg. invalid duration)
- Unit tests are done to make sure all the flags work (testing every flag, and every different config type), and error when they should.
- Unit tests to test merging behaviour is done
- Fixed mergo problem with merging struct pointers (with help from Ben) 
- ConfigDuration now has its duration exported (with help from Ben) 
- logRotationConfig.LocalTime has been made a bool*
- MaxHeartbeat is now pointer for ConfigDuration.
- ConfigDuration have a .Value() getter that defaults to 0 if nil
- legacyconfig.MaxHeartbeat is now *uint instead of just uint
- Added test to make sure number of flags match number of config options that are public

### CODE REVIEW

**Review by bbrks — CHANGES_REQUESTED** [review]
General approach seems reasonable but it looks like it may not be working properly.
It seems to be overwriting my config file options even with no flags set.

```
> cat cnf_bootstrap.json
{
  "bootstrap": {
    "server": "couchbases://localhost",
    "config_update_frequency": "1m",
    "username": "Administrator",
    "password": "password",
    "server_tls_skip_verify": true
  },
  "api": {
    "https": {
      "use_tls_client": false
    }
  },
  "logging": {
    "log_file_path": "/tmp/sglog",
    "console": {
      "log_level": "debug",
      "log_keys": ["HTTP","Config","CRUD","DCP","Sync"]
    }
  }
}
```

```
> go build && ./sync_gateway cnf_bootstrap.json
2021-08-10T15:08:24.557+01:00 ==== Couchbase Sync Gateway/() CE ====
2021-08-10T15:08:24.557+01:00 [INF] Loading content from [cnf_bootstrap.json] ...
2021-08-10T15:08:24.559+01:00 [INF] Config: Starting in persistent mode using config group "default"
2021-08-10T15:08:24.559+01:00 [INF] Logging: Console disabled
2021-08-10T15:08:24.559+01:00 [INF] Logging: Files to /tmp/sglog
^C%
```

...

```
> git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
> go build && ./sync_gateway cnf_bootstrap.json
2021-08-10T15:09:09.569+01:00 ==== Couchbase Sync Gateway/() CE ====
2021-08-10T15:09:09.569+01:00 [INF] Loading content from [cnf_bootstrap.json] ...
2021-08-10T15:09:09.570+01:00 [INF] Config: Starting in persistent mode using config group "default"
2021-08-10T15:09:09.570+01:00 [INF] Logging: Console to stderr
2021-08-10T15:09:09.570+01:00 [INF] Logging: Files to /tmp/sglog
2021-08-10T15:09:09.571+01:00 [INF] Logging: Console level: debug
2021-08-10T15:09:09.571+01:00 [INF] Logging: Console keys: [Config CRUD DCP HTTP Sync]
2021-08-10T15:09:09.571+01:00 [INF] Logging: Redaction level: partial
2021-08-10T15:09:09.571+01:00 [INF] Configured process to allow 5000 open file descriptors
2021-08-10T15:09:09.571+01:00 [INF] Logging stats with frequency: 1m0s
2021-08-10T15:09:13.301+01:00 [INF] Config: Successfully connected to cluster
2021-08-10T15:09:13.421+01:00 [DBG] Config+: "sg_int_0_1628517367577724000" did not contain config in group "default"
2021-08-10T15:09:13.421+01:00 [INF] Config: Updating database "db1" for bucket "b1" with new config from bucket
2021-08-10T15:09:13.421+01:00 [INF] Opening db /db1 as bucket "b1", pool "default", server <couchbases://localhost>
2021-08-10T15:09:13.421+01:00 [INF] GoCBCustomSGTranscoder Opening Couchbase database b1 on <couchbases://localhost> as user "<ud>Administrator</ud>"
2021-08-10T15:09:14.403+01:00 [INF] Successfully opened bucket b1
2021-08-10T15:09:14.424+01:00 [INF] Set query timeouts for bucket b1 to cluster:1m15s, bucket:1m15s
2021-08-10T15:09:14.424+01:00 [INF] Initializing indexes with numReplicas: 0...
2021-08-10T15:09:14.676+01:00 [INF] Verifying index availability for bucket b1...
2021-08-10T15:09:14.684+01:00 [INF] Indexes ready for bucket b1.
2021-08-10T15:09:14.684+01:00 [INF] delta_sync enabled=false with rev_max_age_seconds=86400 for database db1
2021-08-10T15:09:14.686+01:00 [INF] Created background task: "CleanAgedItems" with interval 1m0s
...
```

File: rest/config_flags.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+				var ll base.LogLevel
+				err := ll.UnmarshalText([]byte(*val.flagValue.(*string)))
+				if err != nil {
+					err = fmt.Errorf("flag %s error: %w", f.Name, err)
+					errorMessages = multierror.Append(errorMessages, err)
+					return
+				}
+				rval.Set(reflect.ValueOf(&ll))
+			default:
+				panic(fmt.Sprintf("Error: Unknown type %v for flag %v\n", rval.Type(), f.Name))
```
[the remark this answers is not in the document]
Comment (IsaacLambat) [reply rest/config_flags.go:194]: Leaving as a panic as it should never get to that stage for a user (only a developer when adding new config options but not adding a case for a new/different type). 

Let me know if you still think I should change it to just an error.

File: rest/config.go [hunk 2]
```diff
@@ -794,7 +794,7 @@ func SetMaxFileDescriptors(maxP *uint64) error {
 
 func (sc *ServerContext) Serve(config *StartupConfig, addr string, handler http.Handler) error {
 	http2Enabled := false
-	if config.Unsupported.HTTP2 != nil {
+	if config.Unsupported.HTTP2 != nil && config.Unsupported.HTTP2.Enabled != nil {
```
Comment (bbrks) [inline rest/config.go:830]: 👍

File: rest/config_legacy_test.go [hunk 3]
```diff
@@ -32,8 +32,8 @@ func TestLegacyConfigToStartupConfig(t *testing.T) {
 		{
 			name:     "Override duration zero ServerReadTimeout",
 			base:     StartupConfig{API: APIConfig{ServerReadTimeout: base.NewConfigDuration(time.Second * 10)}},
-			input:    LegacyServerConfig{ServerReadTimeout: base.IntPtr(0)},
-			expected: StartupConfig{API: APIConfig{ServerReadTimeout: base.NewConfigDuration(0)}},
+			input:    LegacyServerConfig{ServerReadTimeout: base.IntPtr(1)},
+			expected: StartupConfig{API: APIConfig{ServerReadTimeout: base.NewConfigDuration(time.Second)}},
```
Comment (bbrks) [inline rest/config_legacy_test.go]: Doesn't match test case name any more - but looking at the test-case again it looks like we'd want a pointer to `0` to override the original value, as it's explicitly set?

I can't remember why this stopped working. Is it related to `D` inside `ConfigDuration` not being a pointer type and so defaulting to zero/empty value?

File: rest/config_flags.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+// configFlag stores the config value, and the corresponding flag value
+type configFlag struct {
+	config    interface{}
+	flagValue interface{}
+}
+
+// registerConfigFlags holds a map of the flag values with the configFlag it goes with
+// (which stores the corresponding config field pointer and flag value).
+func registerConfigFlags(config *StartupConfig, fs *flag.FlagSet) map[string]configFlag {
+	return map[string]configFlag{
```
Comment (bbrks) [inline rest/config_flags.go:23]: I'm wondering how we can catch developer-time errors like adding a new config option but forgetting to update this map, or vice versa.

Is it possible for a test to count all of the fields (including children) in StartupConfig and ensure it matches the number of elements returned by this function? That at least gives us an early-warning if we forget to update flags.

**Review by bbrks — CHANGES_REQUESTED** [review]
The change to the mergo transformer looks good for handling struct pointers, and the general appraoch for flags is fine for now. 

I think we should add a test that ensures flags and StartupConfig are somewhat coupled, to catch scenarios in dev/unit-test time where we add a config option but forget to add a flag, if it's not too difficult to do.

File: rest/changes_api.go [hunk 5]
```diff
@@ -133,12 +133,16 @@ func (h *handler) updateChangesOptionsFromQuery(feed *string, options *db.Change
 	}
 
 	if _, ok := values["heartbeat"]; ok {
+		maxHeartbeat := h.server.config.Replicator.MaxHeartbeat
```
Comment (adamcfraser) [inline rest/changes_api.go]: Can use .Value() here (or below).

**Review by bbrks — DISMISSED** [review]

**Review by bbrks — APPROVED** [review]

**Pull request merged**